allennlp/models/encoder_decoders/composed_seq2seq.py

```diff
@@ -0,0 +1,164 @@
+        value as the source embedder's.
+    beam_size : ``int``, optional (default = None)
+        Width of the beam for beam search. If not specified, greedy decoding is used.
+    scheduled_sampling_ratio : ``float``, optional (default = 0.)
+        At each timestep during training, we sample a random number between 0 and 1, and if it is
+        not less than this value, we use the ground truth labels for the whole batch. Else, we use
+        the predictions from the previous time step for the whole batch. If this value is 0.0
+        (default), this corresponds to teacher forcing, and if it is 1.0, it corresponds to not
+        using target side ground truth labels.  See the following paper for more information:
+        `Scheduled Sampling for Sequence Prediction with Recurrent Neural Networks. Bengio et al.,
+        2015 <https://arxiv.org/abs/1506.03099>`_.
+    tensor_based_metric : ``Metric``, optional (default = BLEU)
+        A metric to track on validation data that takes raw tensors when its called.
+        This metric must accept two arguments when called: a batched tensor
+        of predicted token indices, and a batched tensor of gold token indices.
+    token_based_metric : ``Metric``, optional (default = None)
+        A metric to track on validation data that takes lists of lists of tokens
+        as input. This metric must accept two arguments when called, both
+        of type `List[List[str]]`. The first is a predicted sequence for each item
+        in the batch and the second is a gold sequence for each item in the batch.
+    """
+
+    def __init__(self,
+                 vocab: Vocabulary,
+                 source_embedder: TextFieldEmbedder,
+                 encoder: Seq2SeqEncoder,
+                 decoder_cell: DecoderCell,
+                 max_decoding_steps: int,
+                 beam_size: int = None,
+                 target_namespace: str = "tokens",
+                 target_embedding_dim: int = None,
+                 scheduled_sampling_ratio: float = 0.,
+                 tensor_based_metric: Metric = None,
+                 token_based_metric: Metric = None,
+                 ) -> None:
+
+        super(ComposedSeq2Seq, self).__init__(vocab)
+        self._target_namespace = target_namespace
+        self._scheduled_sampling_ratio = scheduled_sampling_ratio
+
+        # Dense embedding of source vocab tokens.
+        self._source_embedder = source_embedder
+
+        # Encodes the sequence of source embeddings into a sequence of hidden states.
+        self._encoder = encoder
+
+        if self._encoder.get_output_dim() != decoder_cell.get_output_dim():
+            raise ConfigurationError(
+                f"Encoder hidden dimension {self._encoder.get_output_dim()} should be"
+                f" equal to decoder dimension {decoder_cell.get_output_dim()}.")
+
+        self.decoder: SimpleSeqDecoder = SimpleSeqDecoder(
```

Comment:
This should be `Registrable` and a constructor parameter.  Otherwise it's not "fully customizable", as you say in the docstring.

Reply:
Unfortunately I didn't find a way to make this `Registrable`. It depends on `Vocabulary` and as I understood `Vocabulary` is passed implicitly only to `Model` objects.

Reply:
Changed this docstring

Reply:
The vocabulary is implicitly passed everywhere that it's needed - the embedders get access to it, for instance.

Reply:
really so! Made it registrable